microcosm_pubsub/context.py

```diff
@@ -27,43 +26,39 @@ def __init__(self, graph):
         self.enable_ttl = graph.config.sqs_message_context.enable_ttl
         self.initial_ttl = graph.config.sqs_message_context.initial_ttl
 
-    def __call__(self, message, **kwargs):
+    def __call__(self, context, **kwargs) -> Dict[str, str]:
         """
         Create a new context from a message.
 
         """
-        # start with opaque data passed in message
-        if "opaque_data" in message:
-            context = message["opaque_data"].copy()
-        else:
-            context = dict()
-
-        # merge in explicit and derived arguments
-        context.update(
-            **self.uri_for(context, message),
-            **self.ttl_for(context, message),
+        if isinstance(context, SQSMessage):
+            return self.from_sqs_message(context, **kwargs)
+
+        # XXX we want to remove this
+        return self.from_dict(context, **kwargs)
+
+    def from_sqs_message(self, message, **kwargs):
+        context = dict(
+            # include the message id
+            message_id=message.message_id,
+            # include
+            **message.opaque_data,
             **kwargs,
         )
 
-        return context
-
-    def uri_for(self, context, message):
-        uri = message.get("uri")
-        return dict(uri=uri) if uri else dict()
+        # include the TTL (if enabled)
+        if self.enable_ttl:
```

Comment:
I am no longer validating TTL on context load because:
 - Our `elapsed_time` context manager saves its output in `self.opaque`
 - We want our `MessageHandlingResult` to record timing data (for metrics reporting)

So: `self.opaque` needs to be initialized outside of the result invocation (or we can't capture the reporting data).

But:

 - We want our `MessageHandlingResult` to capture all message handling errors (for result reporting)

So: we can't have errors raised in the `self.opaque` initialization